### ISSUE TITLE
Checking for updates should be separate from actually updating

### ISSUE THREAD

**Issue by Podesta** [violentmonkey/violentmonkey#1023]
Violent monkey treats "check for update" as an update authorization, which can lead to serious security risks, on top of being just annoying. Checking for an update does not mean that I'm consenting for a script to be automatically updated without my input or review.

I'm proposing two changes:

1. The check for update should be decoupled from the auto update. I want to be notified if the author of an script updated it, but I still want to make the decision of weather or not I want to update it, after reviewing the changes made. Ideally tied into #500.

2. That is specially the case when you have made local changes to a script. After checking for an update, and giving me the option to actually update the script on a case by case basis, there should be a warning if you have made local changes to a script, so you are reminded of the fact that those changes will be overwritten. A script that I have changed locally should never be updated and overwritten by default.

The way [Stylus](https://github.com/openstyles/stylus) implements is a good example.

**Comment by 720905** [comment]
this is a pretty big deal for me, is there a way you found to make violentmonkey more secure for now?

**Comment by TFWol** [comment]
This, please!

Checking for updates manually by visiting each link to avoid an auto update is daunting with hundreds of scripts.

**Comment by TFWol** [comment]
I still really want this. It even says '**Check** for updates', yet it will absolutely update.

The misleading phrasing should really be changed until this issue is implemented.

**Comment by tophf** [comment]
Someday we'll certainly split it, but note that Firefox also uses "Check for updates" menu command to immediately update the extensions, no questions asked, so it's not like we're being weird.

**Comment by TFWol** [comment]
Well, to be fair that's the browser. Those updates include security fixes. Scripts can be many things.

**Comment by tophf** [comment]
No, my comment is about extensions, which are essentially userscripts, not the browser's security fixes.

**Comment by TFWol** [comment]
Mine doesn't do that. I'm able to pick and choose.

https://github.com/violentmonkey/violentmonkey/assets/9045213/45f09b07-e6ea-4246-a340-42ea3ae6733c

**Comment by tophf** [comment]
I was wrong about Firefox evidently as I used this command a while ago.

Firefox: `Check for updates` + separate installation
Chrome: `Update` + auto-installation

Our behavior currently is a mix of the two, which is indeed weird.

**Comment by tophf** [comment]
Turns out Firefox has a setting in that menu to enable automatic updates, I think it's enabled by default.

**Comment by TFWol** [comment]
Seems so. Hard to tell what was a default since Mozilla keeps making changes to everything 😐

**Comment by tophf** [comment]
I found this particular setting in [Firefox source code](https://searchfox.org/mozilla-central/source/browser/app/profile/firefox.js#51). Judging by the info in the line number gutter it was added in 2010 i.e. it didn't change since. I guess it was added to solve the exact same problem reported here.

**Comment by TFWol** [comment]
Interesting. I also found the setting in my user.js I added to over the years (I prob should have checked that first). 
Near the top, so that was one of the first things I changed so long ago. I guess that's why I've never had to touch that particular setting. Set and forget.

**Comment by yennor** [comment]
I've just added a userscript for amazon.com, and there is no freaking way I would let auto update that. But I would love to see a notification about updates.

**Comment by cyfung1031** [comment]
Hope this can be implemented very soon.

<img width="505" alt="Screen Shot 2023-09-22 at 10 31 34" src="https://github.com/violentmonkey/violentmonkey/assets/44498510/fd39df52-7b2f-4eac-bc6f-f4626e0c155d">

The current behavior is just non-sense. Userscript is not allowed user to edit? WTF.

Just do like Tampermonkey. Notify the user when there is an update to override his own version.
Let user to decide whether they want to update or not when update is available, and at the same time let user to confirm "don't ask me future updates about this script"

**Comment by cyfung1031** [comment]
I think #1289 is a little bit different from this issue.

That is about when the latest script is no longer available. That empty script response shall be considered as invalid not an update.
no really a strong relationship with the notify vs update mechanism mentioned here.

**Comment by cyfung1031** [comment]
And this issue shall be pinned as the readonly userscript is really a breaking change compared with the old version. I think more people will come and complain v2.15.4.

**Comment by tophf** [comment]
Our current behavior is arguably good for 99.9% of users judging by the fact that Chrome automatically updates the extensions without prompting first and extensions are used by hundreds of millions of users over the past decade.

**Comment by tophf** [comment]
As for the read-only protection, the implementation may have bugs, which we will fix before releasing a stable version.
The beta exists to test new features and new code, so it's expected to have problems and bugs.

**Comment by TFWol** [comment]
> arguably good for 99.9% of users judging by the fact that Chrome automatically updates the extensions without prompting first and extensions are used by hundreds of millions of users over the past decade.

Well..the average user doesn't typically install these kinds of addons. Those that don't like how something's currently working do, which means there's a higher chance they also change other things, like the mentioned automatic extension updates.  

I sure do; an update can be poisoned if the developer sells or gets hacked/tricked. Look at what happened with the recent xc liblzma thing.
Also userscripts are so much easier to scrutinize than addons, so they're more likely to want more control over the update process.  😃

**Comment by tesaguri** [comment]
The threat model of browser extensions are quite different from that of user scripts in that they are supposed to be reviewed by the browser vendors before being signed. Also, the vendor can revoke the signature if the extension is found to be malicious.

**Comment by tophf** [comment]
@tesaguri, unfortunately 99% or more of extensions aren't really actually reviewed, the process is mostly automatic. Ideally, one should not trust extensions by default, but everyone does because it makes life easier.

**Comment by TFWol** [comment]
> I just found that "Notify script updates" checker box. I am not sure whether this is the feature I mentioned.

There's a reason I had that disabled. I think I got burned by it somehow. I'll have to figure out what that was when I have energy to test in a clean browser with a few scripts.

> Don't bother tophf. Just let him focus on the MV3 upgrade first 😂

I forget that most people use Chrome
